### PR TITLE
add Subscription.get_active_subscription_by_domain()

### PR DESCRIPTION
**Pull request by nickpell**
This is the second half to https://github.com/dimagi/commcare-hq/pull/15852 (wanted to wait one invoicing period to make sure nothing unexpectedly broke).

Since we can now know that there is at most one active subscription per domain, we're able to clean up the query to find a domain's active subscription (https://github.com/dimagi/commcare-hq/commit/39188fd4f0b3e45922649b67365bfb1485981b42).

@orangejenny

### CODE REVIEW

File: corehq/apps/accounting/models.py [hunk 1]
```diff
@@ -1558,43 +1558,34 @@ def set_billing_account_entry_point(self):
...
+                return DefaultProductPlan.get_default_plan_version()
             except DefaultProductPlan.DoesNotExist:
                 raise ProductPlanNotFoundError
-        domain = domain_obj
-        subscriber = Subscriber.objects.safe_get(domain=domain.name)
-        plan_version, subscription = cls._get_plan_by_subscriber(subscriber) if subscriber else (None, None)
-        if plan_version is None:
-            plan_version = DefaultProductPlan.get_default_plan_version()
-        return plan_version, subscription
+        else:
```
Comment (orangejenny) [inline corehq/apps/accounting/models.py]: nitpick, but this doesn't need to be inside an `else`

File: corehq/apps/accounting/models.py [hunk 2]
```diff
@@ -1558,43 +1558,34 @@ def set_billing_account_entry_point(self):
...
+            active_subscription = cls.get_active_subscription_by_domain(domain_obj.name)
+            if active_subscription is not None:
+                return active_subscription.plan_version
+            else:
```
Comment (orangejenny) [inline corehq/apps/accounting/models.py]: same here

File: corehq/apps/api/domain_metadata.py [hunk 3]
```diff
@@ -47,15 +47,13 @@ def dispatch(self, request_type, request, **kwargs):
...
-            Subscription.get_subscribed_plan_by_domain(domain)
-        )
+        subscription = Subscription.get_active_subscription_by_domain(domain.name)
         return {
             "date_start": (subscription.date_start
                            if subscription is not None else None),
             "date_end": (subscription.date_end
                          if subscription is not None else None),
-            "plan_version": plan_version,
+            "plan_version": subscription.plan_version,
```
Comment (orangejenny) [inline corehq/apps/api/domain_metadata.py]: is it guaranteed that `subscription` won't be `None` here?

File: corehq/apps/domain/views.py [hunk 4]
```diff
@@ -677,7 +677,8 @@ def can_purchase_credits(self):
     @property
     @memoized
     def plan(self):
-        plan_version, subscription = Subscription.get_subscribed_plan_by_domain(self.domain_object)
+        subscription = Subscription.get_active_subscription_by_domain(self.domain)
+        plan_version = subscription.plan_version
```
Comment (orangejenny) [inline corehq/apps/domain/views.py]: This should probably check for `subscription` being `None` (there are similar checks in this function)

File: corehq/apps/accounting/models.py [hunk 5]
```diff
@@ -1558,43 +1558,34 @@ def set_billing_account_entry_point(self):
...
-        current_subscription = active_subscriptions[0]
-        return current_subscription.plan_version, current_subscription
+        except cls.DoesNotExist:
+            return None
 
     @classmethod
     def get_subscribed_plan_by_domain(cls, domain):
         """
-        Returns SoftwarePlanVersion, Subscription for the given domain.
+        Returns SoftwarePlanVersion for the given domain.
```
Comment (orangejenny) [inline corehq/apps/accounting/models.py]: so much nicer!

File: corehq/apps/userreports/views.py [hunk 6]
```diff
@@ -301,8 +301,7 @@ def page_context(self):
     @property
     @memoized
     def plan_name(self):
-        plan_version, _ = Subscription.get_subscribed_plan_by_domain(self.domain)
-        return plan_version.plan.name
+        return Subscription.get_subscribed_plan_by_domain(self.domain).plan.name
```
Comment (orangejenny) [inline corehq/apps/userreports/views.py]: should this check for `None`?

**Review by orangejenny — CHANGES_REQUESTED** [review]
There are a couple of places where it looks like you should check for `None`. Otherwise looks good.

File: corehq/apps/api/domain_metadata.py (hunk 3 above)
Comment (nickpell) [inline corehq/apps/api/domain_metadata.py]: updated

File: corehq/apps/domain/views.py (hunk 4 above)
Comment (nickpell) [inline corehq/apps/domain/views.py]: updated

File: corehq/apps/userreports/views.py (hunk 6 above)
Comment (nickpell) [inline corehq/apps/userreports/views.py]: ```get_subscribed_plan_by_domain``` won't return ```None```, so this one is all set

**Comment by nickpell** [comment]
updated ```None``` checks

**Comment by orangejenny** [comment]
New code looks good, but test failure looks real:
```
File "/mnt/commcare-hq-ro/corehq/apps/users/util.py", line 175, in can_add_extra_mobile_workers
    user_limit = request.plan.user_limit
AttributeError: 'WSGIRequest' object has no attribute 'plan'```

**Comment by nickpell** [comment]
✅

**Review by orangejenny — APPROVED** [review]

**Pull request merged**